### DISCUSSION TITLE
Internal procedures are not invoked or referenced in the app

### DISCUSSION THREAD

**Discussion by rvanbekkum** [StefanMaron/BusinessCentral.LinterCop#197] [New Rule]
Through this discussion I would like to suggest a new rule:

If the `internalsVisibleTo` setting is _not_ set or set to an empty array (`[]`) in `app.json`, then the rule will raise a warning for procedures that are not used/invoked anywhere in the app.
This would be useful to avoid dangling procedures in the app.

So, it would be a rule similar to rule [AA0228](https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/analyzers/codecop-aa0228), "The method '{0}' is declared but never used.".

**Comment by rvanbekkum** [comment]
Should take into account internal procedures that are used for notifications (i.e., with a `Notification` parameter), e.g.:

```al
internal procedure InvokeActionFromNotification(HostNotification: Notification)
begin
    // Implementation here
end;
```

P.S. I have already implemented this rule in an internal code analyzer, because we needed it. We are first trying this implementation out internally and might make a PR to the LinterCop later on.

**Reply by tinestaric** [reply]
@rvanbekkum 
Awesome! This is exactly what I was planning on implementing with our internal code analyzer as well but never found the time. Would you mind sharing your implementation? We could help polish/test it out if you think it's not ready for the BC.LinterCop.

**Reply by rvanbekkum** [reply]
Hi @tinestaric,
I will check for maybe making a PR to the BC.LinterCop repository.

One thing that could be improved about our implementation is that it just skips internal procedures with a single `Notification` parameter, just making the assumption they are used for notification actions (which is a quite fair assumption, which is true in 99% of the cases, but still, not necessarily _always_ true).

**Reply by rvanbekkum** [reply]
Considering no PRs are actively reviewed or merged, I have actually never been motivated to make a PR. 😅
In the mean-time we did improve the implemented code analyzer diagnostics which should cover this:

| ID     | Description                                                                                                    | Additional Info                                                                                                                                                                                        | Default Severity |
|--------|----------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------|
| WS0003 | The internal method is declared but never used.                                                                | Also detects unused public procedures in objects with Access = Internal.                                                                                                                               | Warning          |
| WS0020 | The internal method is only used in the object in which it is declared. Consider making the procedure local.    |            | Info             |

Waiting out for PR reviews to be picked up again at the moment. 🙂

**Reply by Arthurvdv** [reply]
> Waiting out for PR reviews to be picked up again at the moment. 🙂

I would love to review a PR 😉

**Reply by rvanbekkum** [reply]
I'll check for setting up a PR.

**Reply by fvet** [reply]
> Should take into account internal procedures that are used for notifications (i.e., with a Notification parameter), e.g.:

Please consider the internal functions with "ErrorInfo" parameters as well. These are used by error with an action ('Fix It' pattern), similar to notifications.

**Reply by rvanbekkum** [reply]
> > Should take into account internal procedures that are used for notifications (i.e., with a Notification parameter), e.g.:
> 
> Please consider the internal functions with "ErrorInfo" parameters as well. These are used by error with an action ('Fix It' pattern), similar to notifications.

Funny, I ran into this last week, so adjusted the codecop rule for it. 😊 If more special cases come to mind, please let me know.

**Reply by rvanbekkum** [reply]
> > Waiting out for PR reviews to be picked up again at the moment. 🙂
> 
> I would love to review a PR 😉

@Arthurvdv I opened a PR for review: https://github.com/StefanMaron/BusinessCentral.LinterCop/pull/514

**Reply by Arthurvdv** [reply]
Thank you for creating a PR, very much appreciated! I'll do my best to review this week.

**Reply by Arthurvdv** [reply]
> Default severity for now set to Info for both rules even though it is advised to have LC0052 as a warning.

Next to [cache invalidation and naming things](https://martinfowler.com/bliki/TwoHardThings.html), is determine the right Default severity the third hard thing here 😅

I've learned from @StefanMaron to be reluctant to introduce new rules with a default severity set to warning, due to possibility of breaking existing pipelines out there. I agree on you here, and at the same time I'm glad you set both of them to info where these can be easily raised to a warning with a custom ruleset.

Again thank you for making a PR for these, I really like these rules and probably will set these as warnings in our pipelines.
